{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "5158af5e-c4e5-440b-b64a-c176608d4960",
   "metadata": {},
   "source": [
    "# Análisis de enriquecimiento con blitzGSEA\n",
    "```\n",
    "Autor:  Martín Sende Pombo (email: martinsendepombo@outlook.com)\n",
    "Se utilizó ChatGPT 3.5 como asistente de programación, para elaborar este código basado en los ejemplos proporcionados por el Ma'ayan Laboratory.\n",
    "Creado: 16-01-2024\n",
    "Última modificación: 13-08-2024\n",
    "Propósito: Este cuaderno de Jupyter permite efectuar un análisis de enriquecimiento de conjuntos de genes (GSEA) utilizando la biblioteca blitzGSEA.\n",
    "```\n",
    "Copyright (C) 2024  Martín Sende Pombo\n",
    "\n",
    "    Este programa es software libre: puede redistribuirlo y/o modificarlo\n",
    "    bajo los términos de la Licencia Pública General de GNU publicada por la\n",
    "    Free Software Foundation, ya sea la versión 3 de la Licencia,\n",
    "    o (a su elección) cualquier versión posterior.\n",
    "\n",
    "    Este programa se distribuye con la esperanza de que sea útil,\n",
    "    pero SIN NINGUNA GARANTÍA; ni siquiera la garantía implícita\n",
    "    de COMERCIABILIDAD o IDONEIDAD PARA UN PROPÓSITO PARTICULAR.\n",
    "    Consulte la Licencia Pública General GNU para más detalles.\n",
    "\n",
    "    Debería haber recibido una copia de la Licencia Pública General\n",
    "    GNU junto con este programa. Si no es así, consulte <https://www.gnu.org/licenses/>."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "59b31757-40c3-4249-ab91-3578706aaa4e",
   "metadata": {},
   "source": [
    "## Instalación"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "296f4800-dec8-402b-b770-672a456d06d1",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: blitzgsea in c:\\users\\martin\\anaconda3\\lib\\site-packages (1.3.40)\n",
      "Requirement already satisfied: pandas>=1.1.5 in c:\\users\\martin\\anaconda3\\lib\\site-packages (from blitzgsea) (2.0.3)\n",
      "Requirement already satisfied: numpy in c:\\users\\martin\\anaconda3\\lib\\site-packages (from blitzgsea) (1.24.3)\n",
      "Requirement already satisfied: scikit-learn in c:\\users\\martin\\anaconda3\\lib\\site-packages (from blitzgsea) (1.3.0)\n",
      "Requirement already satisfied: tqdm in c:\\users\\martin\\anaconda3\\lib\\site-packages (from blitzgsea) (4.65.0)\n",
      "Requirement already satisfied: statsmodels in c:\\users\\martin\\anaconda3\\lib\\site-packages (from blitzgsea) (0.14.0)\n",
      "Requirement already satisfied: mpmath in c:\\users\\martin\\anaconda3\\lib\\site-packages (from blitzgsea) (1.3.0)\n",
      "Requirement already satisfied: python-dateutil>=2.8.2 in c:\\users\\martin\\anaconda3\\lib\\site-packages (from pandas>=1.1.5->blitzgsea) (2.8.2)\n",
      "Requirement already satisfied: pytz>=2020.1 in c:\\users\\martin\\anaconda3\\lib\\site-packages (from pandas>=1.1.5->blitzgsea) (2023.3.post1)\n",
      "Requirement already satisfied: tzdata>=2022.1 in c:\\users\\martin\\anaconda3\\lib\\site-packages (from pandas>=1.1.5->blitzgsea) (2023.3)\n",
      "Requirement already satisfied: scipy>=1.5.0 in c:\\users\\martin\\anaconda3\\lib\\site-packages (from scikit-learn->blitzgsea) (1.11.1)\n",
      "Requirement already satisfied: joblib>=1.1.1 in c:\\users\\martin\\anaconda3\\lib\\site-packages (from scikit-learn->blitzgsea) (1.2.0)\n",
      "Requirement already satisfied: threadpoolctl>=2.0.0 in c:\\users\\martin\\anaconda3\\lib\\site-packages (from scikit-learn->blitzgsea) (2.2.0)\n",
      "Requirement already satisfied: patsy>=0.5.2 in c:\\users\\martin\\anaconda3\\lib\\site-packages (from statsmodels->blitzgsea) (0.5.3)\n",
      "Requirement already satisfied: packaging>=21.3 in c:\\users\\martin\\anaconda3\\lib\\site-packages (from statsmodels->blitzgsea) (23.1)\n",
      "Requirement already satisfied: colorama in c:\\users\\martin\\anaconda3\\lib\\site-packages (from tqdm->blitzgsea) (0.4.6)\n",
      "Requirement already satisfied: six in c:\\users\\martin\\anaconda3\\lib\\site-packages (from patsy>=0.5.2->statsmodels->blitzgsea) (1.16.0)\n"
     ]
    }
   ],
   "source": [
    "!pip3 install blitzgsea"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6fcbd063-b27f-451c-a6a0-e01d33431b68",
   "metadata": {},
   "source": [
    "## Ejecución del análisis de enriquecimiento con blitzGSEA"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7ea10742-bbde-4312-8a51-409459ba6d1d",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "### Ejemplo en Python"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "3cd18d05-42ea-4f11-a44d-6d070eed0c9c",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'LIBRARY_LIST_URL': 'https://maayanlab.cloud/speedrichr/api/listlibs', 'LIBRARY_DOWNLOAD_URL': 'https://maayanlab.cloud/Enrichr/geneSetLibrary?mode=text&libraryName='}\n",
      "0 - GeneSigDB\n",
      "1 - Enrichr_Submissions_TF-Gene_Coocurrence\n",
      "2 - SysMyo_Muscle_Gene_Sets\n",
      "3 - WikiPathway_2021_Human\n",
      "4 - HomoloGene\n",
      "5 - WikiPathways_2013\n",
      "6 - PFOCR_Pathways_2023\n",
      "7 - serine_threonine_kinome_atlas_2023\n",
      "8 - OMIM_Disease\n",
      "9 - Data_Acquisition_Method_Most_Popular_Genes\n",
      "10 - Cancer_Cell_Line_Encyclopedia\n",
      "11 - WikiPathways_2016\n",
      "12 - WikiPathways_2015\n",
      "13 - RNAseq_Automatic_GEO_Signatures_Human_Up\n",
      "14 - Human_Gene_Atlas\n",
      "15 - KOMP2_Mouse_Phenotypes_2022\n",
      "16 - MoTrPAC_2023\n",
      "17 - Kinase_Perturbations_from_GEO_down\n",
      "18 - Disease_Signatures_from_GEO_down_2014\n",
      "19 - Disease_Perturbations_from_GEO_up\n",
      "20 - Old_CMAP_down\n",
      "21 - MCF7_Perturbations_from_GEO_up\n",
      "22 - NIH_Funded_PIs_2017_GeneRIF_ARCHS4_Predictions\n",
      "23 - PPI_Hub_Proteins\n",
      "24 - DepMap_WG_CRISPR_Screens_Sanger_CellLines_2019\n",
      "25 - Disease_Signatures_from_GEO_up_2014\n",
      "26 - GTEx_Tissue_Expression_Up\n",
      "27 - L1000_Kinase_and_GPCR_Perturbations_up\n",
      "28 - ARCHS4_Cell-lines\n",
      "29 - VirusMINT\n",
      "30 - KEGG_2019_Human\n",
      "31 - ARCHS4_Tissues\n",
      "32 - MGI_Mammalian_Phenotype_Level_4\n",
      "33 - The_Kinase_Library_2023\n",
      "34 - MGI_Mammalian_Phenotype_Level_3\n",
      "35 - InterPro_Domains_2019\n",
      "36 - KEGG_2015\n",
      "37 - MSigDB_Computational\n",
      "38 - KEGG_2013\n",
      "39 - TF-LOF_Expression_from_GEO\n",
      "40 - GWAS_Catalog_2019\n",
      "41 - KEGG_2016\n",
      "42 - NCI-Nature_2015\n",
      "43 - NCI-Nature_2016\n",
      "44 - CCLE_Proteomics_2020\n",
      "45 - PheWeb_2019\n",
      "46 - GeDiPNet_2023\n",
      "47 - RNA-Seq_Disease_Gene_and_Drug_Signatures_from_GEO\n",
      "48 - LINCS_L1000_Chem_Pert_down\n",
      "49 - Old_CMAP_up\n",
      "50 - LINCS_L1000_Ligand_Perturbations_down\n",
      "51 - Enrichr_Users_Contributed_Lists_2020\n",
      "52 - NIH_Funded_PIs_2017_Human_GeneRIF\n",
      "53 - Jensen_TISSUES\n",
      "54 - Azimuth_Cell_Types_2021\n",
      "55 - DisGeNET\n",
      "56 - Panther_2016\n",
      "57 - LINCS_L1000_Ligand_Perturbations_up\n",
      "58 - Rare_Diseases_AutoRIF_Gene_Lists\n",
      "59 - TargetScan_microRNA\n",
      "60 - Achilles_fitness_increase\n",
      "61 - Panther_2015\n",
      "62 - WikiPathways_2019_Mouse\n",
      "63 - ARCHS4_TFs_Coexp\n",
      "64 - LINCS_L1000_Chem_Pert_up\n",
      "65 - MSigDB_Oncogenic_Signatures\n",
      "66 - Gene_Perturbations_from_GEO_down\n",
      "67 - Table_Mining_of_CRISPR_Studies\n",
      "68 - Rare_Diseases_GeneRIF_Gene_Lists\n",
      "69 - Ligand_Perturbations_from_GEO_down\n",
      "70 - SILAC_Phosphoproteomics\n",
      "71 - Ligand_Perturbations_from_GEO_up\n",
      "72 - Drug_Perturbations_from_GEO_up\n",
      "73 - SynGO_2022\n",
      "74 - Allen_Brain_Atlas_10x_scRNA_2021\n",
      "75 - MGI_Mammalian_Phenotype_Level_4_2021\n",
      "76 - ClinVar_2019\n",
      "77 - GWAS_Catalog_2023\n",
      "78 - MAGMA_Drugs_and_Diseases\n",
      "79 - KEA_2015\n",
      "80 - KEA_2013\n",
      "81 - Microbe_Perturbations_from_GEO_up\n",
      "82 - Chromosome_Location\n",
      "83 - COVID-19_Related_Gene_Sets_2021\n",
      "84 - MGI_Mammalian_Phenotype_Level_4_2019\n",
      "85 - ARCHS4_IDG_Coexp\n",
      "86 - NIH_Funded_PIs_2017_AutoRIF_ARCHS4_Predictions\n",
      "87 - lncHUB_lncRNA_Co-Expression\n",
      "88 - DrugMatrix\n",
      "89 - Virus_Perturbations_from_GEO_down\n",
      "90 - huMAP\n",
      "91 - L1000_Kinase_and_GPCR_Perturbations_down\n",
      "92 - Elsevier_Pathway_Collection\n",
      "93 - NIH_Funded_PIs_2017_Human_AutoRIF\n",
      "94 - Diabetes_Perturbations_GEO_2022\n",
      "95 - RNAseq_Automatic_GEO_Signatures_Mouse_Down\n",
      "96 - ESCAPE\n",
      "97 - UK_Biobank_GWAS_v1\n",
      "98 - Aging_Perturbations_from_GEO_up\n",
      "99 - Human_Phenotype_Ontology\n",
      "100 - Proteomics_Drug_Atlas_2023\n",
      "101 - dbGaP\n",
      "102 - SubCell_BarCode\n",
      "103 - Transcription_Factor_PPIs\n",
      "104 - GO_Cellular_Component_2017b\n",
      "105 - HuBMAP_ASCTplusB_augmented_2022\n",
      "106 - MSigDB_Hallmark_2020\n",
      "107 - GlyGen_Glycosylated_Proteins_2022\n",
      "108 - MAGNET_2023\n",
      "109 - BioPlanet_2019\n",
      "110 - HDSigDB_Human_2021\n",
      "111 - GTEx_Tissues_V8_2023\n",
      "112 - GTEx_Tissue_Expression_Down\n",
      "113 - Metabolomics_Workbench_Metabolites_2022\n",
      "114 - Tissue_Protein_Expression_from_Human_Proteome_Map\n",
      "115 - Epigenomics_Roadmap_HM_ChIP-seq\n",
      "116 - PhenGenI_Association_2021\n",
      "117 - MCF7_Perturbations_from_GEO_down\n",
      "118 - ProteomicsDB_2020\n",
      "119 - Virus-Host_PPI_P-HIPSTer_2020\n",
      "120 - OMIM_Expanded\n",
      "121 - Reactome_2022\n",
      "122 - Genes_Associated_with_NIH_Grants\n",
      "123 - CellMarker_Augmented_2021\n",
      "124 - ENCODE_and_ChEA_Consensus_TFs_from_ChIP-X\n",
      "125 - HDSigDB_Mouse_2021\n",
      "126 - Jensen_COMPARTMENTS\n",
      "127 - ChEA_2015\n",
      "128 - ChEA_2016\n",
      "129 - KEGG_2019_Mouse\n",
      "130 - ChEA_2022\n",
      "131 - Drug_Perturbations_from_GEO_2014\n",
      "132 - LINCS_L1000_Chem_Pert_Consensus_Sigs\n",
      "133 - TargetScan_microRNA_2017\n",
      "134 - KEGG_2021_Mouse\n",
      "135 - Allen_Brain_Atlas_down\n",
      "136 - BioCarta_2013\n",
      "137 - WikiPathways_2019_Human\n",
      "138 - Reactome_2013\n",
      "139 - Rummagene_transcription_factors\n",
      "140 - Gene_Perturbations_from_GEO_up\n",
      "141 - GO_Cellular_Component_2015\n",
      "142 - Rummagene-signatures\n",
      "143 - GO_Cellular_Component_2013\n",
      "144 - BioCarta_2016\n",
      "145 - BioCarta_2015\n",
      "146 - Reactome_2015\n",
      "147 - GO_Cellular_Component_2018\n",
      "148 - Reactome_2016\n",
      "149 - GO_Cellular_Component_2017\n",
      "150 - GO_Cellular_Component_2023\n",
      "151 - GO_Cellular_Component_2021\n",
      "152 - WikiPathway_2021_Mouse\n",
      "153 - ENCODE_TF_ChIP-seq_2015\n",
      "154 - ENCODE_TF_ChIP-seq_2014\n",
      "155 - RNAseq_Automatic_GEO_Signatures_Mouse_Up\n",
      "156 - GO_Molecular_Function_2017b\n",
      "157 - FANTOM6_lncRNA_KD_DEGs\n",
      "158 - MGI_Mammalian_Phenotype_2013\n",
      "159 - NCI-60_Cancer_Cell_Lines\n",
      "160 - HMS_LINCS_KinomeScan\n",
      "161 - Azimuth_2023\n",
      "162 - MGI_Mammalian_Phenotype_2017\n",
      "163 - Rare_Diseases_GeneRIF_ARCHS4_Predictions\n",
      "164 - Virus_Perturbations_from_GEO_up\n",
      "165 - SigComLINCS_Drug_Consensus_Signatures\n",
      "166 - PFOCR_Pathways\n",
      "167 - Enrichr_Libraries_Most_Popular_Genes\n",
      "168 - IDG_Drug_Targets_2022\n",
      "169 - Orphanet_Augmented_2021\n",
      "170 - GO_Biological_Process_2021\n",
      "171 - TRANSFAC_and_JASPAR_PWMs\n",
      "172 - GO_Biological_Process_2023\n",
      "173 - COVID-19_Related_Gene_Sets\n",
      "174 - Rare_Diseases_AutoRIF_ARCHS4_Predictions\n",
      "175 - Kinase_Perturbations_from_GEO_up\n",
      "176 - Descartes_Cell_Types_and_Tissue_2021\n",
      "177 - Tabula_Sapiens\n",
      "178 - Tabula_Muris\n",
      "179 - TF_Perturbations_Followed_by_Expression\n",
      "180 - Rummagene_kinases\n",
      "181 - GTEx_Aging_Signatures_2021\n",
      "182 - Tissue_Protein_Expression_from_ProteomicsDB\n",
      "183 - Aging_Perturbations_from_GEO_down\n",
      "184 - DepMap_CRISPR_GeneDependency_CellLines_2023\n",
      "185 - GO_Biological_Process_2013\n",
      "186 - GO_Biological_Process_2017b\n",
      "187 - GO_Biological_Process_2018\n",
      "188 - CORUM\n",
      "189 - GO_Biological_Process_2015\n",
      "190 - Phosphatase_Substrates_from_DEPOD\n",
      "191 - BioPlex_2017\n",
      "192 - TRRUST_Transcription_Factors_2019\n",
      "193 - GO_Biological_Process_2017\n",
      "194 - Pfam_InterPro_Domains\n",
      "195 - HuBMAP_ASCT_plus_B_augmented_w_RNAseq_Coexpression\n",
      "196 - Pfam_Domains_2019\n",
      "197 - WikiPathway_2023_Human\n",
      "198 - Allen_Brain_Atlas_up\n",
      "199 - NURSA_Human_Endogenous_Complexome\n",
      "200 - Genome_Browser_PWMs\n",
      "201 - HumanCyc_2015\n",
      "202 - HumanCyc_2016\n",
      "203 - Rummagene_signatures\n",
      "204 - Chromosome_Location_hg19\n",
      "205 - ChEA_2013\n",
      "206 - Mouse_Gene_Atlas\n",
      "207 - miRTarBase_2017\n",
      "208 - GO_Molecular_Function_2023\n",
      "209 - Jensen_DISEASES\n",
      "210 - RNAseq_Automatic_GEO_Signatures_Human_Down\n",
      "211 - Rummagene-transcription-factors\n",
      "212 - ARCHS4_Kinases_Coexp\n",
      "213 - Microbe_Perturbations_from_GEO_down\n",
      "214 - PanglaoDB_Augmented_2021\n",
      "215 - ENCODE_Histone_Modifications_2013\n",
      "216 - ENCODE_Histone_Modifications_2015\n",
      "217 - Achilles_fitness_decrease\n",
      "218 - DSigDB\n",
      "219 - Disease_Perturbations_from_GEO_down\n",
      "220 - DepMap_WG_CRISPR_Screens_Broad_CellLines_2019\n",
      "221 - Drug_Perturbations_from_GEO_down\n",
      "222 - GO_Molecular_Function_2021\n",
      "223 - GO_Molecular_Function_2017\n",
      "224 - GO_Molecular_Function_2018\n",
      "225 - GO_Molecular_Function_2013\n",
      "226 - GO_Molecular_Function_2015\n",
      "227 - Rummagene-kinases\n",
      "228 - TG_GATES_2020\n",
      "229 - KEGG_2021_Human\n",
      "230 - HMDB_Metabolites\n",
      "231 - LINCS_L1000_CRISPR_KO_Consensus_Sigs\n"
     ]
    }
   ],
   "source": [
    "import blitzgsea as blitz\n",
    "import pandas as pd\n",
    "\n",
    "# read signature as pandas dataframe\n",
    "signature = pd.read_csv(\"https://github.com/MaayanLab/blitzgsea/raw/main/testing/ageing_muscle_gtex.tsv\")\n",
    "\n",
    "# list available gene set libraries in Enrichr\n",
    "blitz.enrichr.print_libraries()\n",
    "\n",
    "# use enrichr submodule to retrieve gene set library\n",
    "library = blitz.enrichr.get_library(\"KEGG_2021_Human\")\n",
    "\n",
    "# run enrichment analysis\n",
    "if __name__ == \"__main__\":  # make sure process is main, when run in a script it can cause errors otherwise\n",
    "  result = blitz.gsea(signature, library)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "daf8f676-4286-4b97-9124-1cba67fcdd5f",
   "metadata": {},
   "source": [
    "### Bibliotecas de conjuntos de genes seleccionables:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "4df7c45c-86b3-40fb-9338-86f87d9c4ff1",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'LIBRARY_LIST_URL': 'https://maayanlab.cloud/speedrichr/api/listlibs', 'LIBRARY_DOWNLOAD_URL': 'https://maayanlab.cloud/Enrichr/geneSetLibrary?mode=text&libraryName='}\n",
      "0 - GeneSigDB\n",
      "1 - Enrichr_Submissions_TF-Gene_Coocurrence\n",
      "2 - SysMyo_Muscle_Gene_Sets\n",
      "3 - WikiPathway_2021_Human\n",
      "4 - HomoloGene\n",
      "5 - WikiPathways_2013\n",
      "6 - PFOCR_Pathways_2023\n",
      "7 - serine_threonine_kinome_atlas_2023\n",
      "8 - OMIM_Disease\n",
      "9 - Data_Acquisition_Method_Most_Popular_Genes\n",
      "10 - Cancer_Cell_Line_Encyclopedia\n",
      "11 - WikiPathways_2016\n",
      "12 - WikiPathways_2015\n",
      "13 - RNAseq_Automatic_GEO_Signatures_Human_Up\n",
      "14 - Human_Gene_Atlas\n",
      "15 - KOMP2_Mouse_Phenotypes_2022\n",
      "16 - MoTrPAC_2023\n",
      "17 - Kinase_Perturbations_from_GEO_down\n",
      "18 - Disease_Signatures_from_GEO_down_2014\n",
      "19 - Disease_Perturbations_from_GEO_up\n",
      "20 - Old_CMAP_down\n",
      "21 - MCF7_Perturbations_from_GEO_up\n",
      "22 - NIH_Funded_PIs_2017_GeneRIF_ARCHS4_Predictions\n",
      "23 - PPI_Hub_Proteins\n",
      "24 - DepMap_WG_CRISPR_Screens_Sanger_CellLines_2019\n",
      "25 - Disease_Signatures_from_GEO_up_2014\n",
      "26 - GTEx_Tissue_Expression_Up\n",
      "27 - L1000_Kinase_and_GPCR_Perturbations_up\n",
      "28 - ARCHS4_Cell-lines\n",
      "29 - VirusMINT\n",
      "30 - KEGG_2019_Human\n",
      "31 - ARCHS4_Tissues\n",
      "32 - MGI_Mammalian_Phenotype_Level_4\n",
      "33 - The_Kinase_Library_2023\n",
      "34 - MGI_Mammalian_Phenotype_Level_3\n",
      "35 - InterPro_Domains_2019\n",
      "36 - KEGG_2015\n",
      "37 - MSigDB_Computational\n",
      "38 - TF-LOF_Expression_from_GEO\n",
      "39 - KEGG_2013\n",
      "40 - GWAS_Catalog_2019\n",
      "41 - KEGG_2016\n",
      "42 - NCI-Nature_2015\n",
      "43 - NCI-Nature_2016\n",
      "44 - CCLE_Proteomics_2020\n",
      "45 - PheWeb_2019\n",
      "46 - GeDiPNet_2023\n",
      "47 - RNA-Seq_Disease_Gene_and_Drug_Signatures_from_GEO\n",
      "48 - LINCS_L1000_Chem_Pert_down\n",
      "49 - Old_CMAP_up\n",
      "50 - LINCS_L1000_Ligand_Perturbations_down\n",
      "51 - Enrichr_Users_Contributed_Lists_2020\n",
      "52 - NIH_Funded_PIs_2017_Human_GeneRIF\n",
      "53 - Jensen_TISSUES\n",
      "54 - Azimuth_Cell_Types_2021\n",
      "55 - LINCS_L1000_Ligand_Perturbations_up\n",
      "56 - DisGeNET\n",
      "57 - Panther_2016\n",
      "58 - Rare_Diseases_AutoRIF_Gene_Lists\n",
      "59 - TargetScan_microRNA\n",
      "60 - Achilles_fitness_increase\n",
      "61 - Panther_2015\n",
      "62 - WikiPathways_2019_Mouse\n",
      "63 - ARCHS4_TFs_Coexp\n",
      "64 - LINCS_L1000_Chem_Pert_up\n",
      "65 - MSigDB_Oncogenic_Signatures\n",
      "66 - Gene_Perturbations_from_GEO_down\n",
      "67 - Table_Mining_of_CRISPR_Studies\n",
      "68 - Rare_Diseases_GeneRIF_Gene_Lists\n",
      "69 - Ligand_Perturbations_from_GEO_down\n",
      "70 - SILAC_Phosphoproteomics\n",
      "71 - Ligand_Perturbations_from_GEO_up\n",
      "72 - Drug_Perturbations_from_GEO_up\n",
      "73 - SynGO_2022\n",
      "74 - SynGO_2024\n",
      "75 - Allen_Brain_Atlas_10x_scRNA_2021\n",
      "76 - MGI_Mammalian_Phenotype_Level_4_2021\n",
      "77 - ClinVar_2019\n",
      "78 - GWAS_Catalog_2023\n",
      "79 - MAGMA_Drugs_and_Diseases\n",
      "80 - KEA_2015\n",
      "81 - KEA_2013\n",
      "82 - Microbe_Perturbations_from_GEO_up\n",
      "83 - Chromosome_Location\n",
      "84 - COVID-19_Related_Gene_Sets_2021\n",
      "85 - MGI_Mammalian_Phenotype_Level_4_2019\n",
      "86 - ARCHS4_IDG_Coexp\n",
      "87 - NIH_Funded_PIs_2017_AutoRIF_ARCHS4_Predictions\n",
      "88 - lncHUB_lncRNA_Co-Expression\n",
      "89 - DrugMatrix\n",
      "90 - Virus_Perturbations_from_GEO_down\n",
      "91 - huMAP\n",
      "92 - L1000_Kinase_and_GPCR_Perturbations_down\n",
      "93 - Elsevier_Pathway_Collection\n",
      "94 - Diabetes_Perturbations_GEO_2022\n",
      "95 - NIH_Funded_PIs_2017_Human_AutoRIF\n",
      "96 - RNAseq_Automatic_GEO_Signatures_Mouse_Down\n",
      "97 - ESCAPE\n",
      "98 - UK_Biobank_GWAS_v1\n",
      "99 - Aging_Perturbations_from_GEO_up\n",
      "100 - Human_Phenotype_Ontology\n",
      "101 - Proteomics_Drug_Atlas_2023\n",
      "102 - dbGaP\n",
      "103 - SubCell_BarCode\n",
      "104 - Transcription_Factor_PPIs\n",
      "105 - GO_Cellular_Component_2017b\n",
      "106 - HuBMAP_ASCTplusB_augmented_2022\n",
      "107 - MSigDB_Hallmark_2020\n",
      "108 - GlyGen_Glycosylated_Proteins_2022\n",
      "109 - MAGNET_2023\n",
      "110 - CellMarker_2024\n",
      "111 - BioPlanet_2019\n",
      "112 - HDSigDB_Human_2021\n",
      "113 - GTEx_Tissues_V8_2023\n",
      "114 - GTEx_Tissue_Expression_Down\n",
      "115 - Metabolomics_Workbench_Metabolites_2022\n",
      "116 - Tissue_Protein_Expression_from_Human_Proteome_Map\n",
      "117 - Epigenomics_Roadmap_HM_ChIP-seq\n",
      "118 - PhenGenI_Association_2021\n",
      "119 - MCF7_Perturbations_from_GEO_down\n",
      "120 - ProteomicsDB_2020\n",
      "121 - Virus-Host_PPI_P-HIPSTer_2020\n",
      "122 - OMIM_Expanded\n",
      "123 - Reactome_2022\n",
      "124 - CellMarker_Augmented_2021\n",
      "125 - Genes_Associated_with_NIH_Grants\n",
      "126 - ENCODE_and_ChEA_Consensus_TFs_from_ChIP-X\n",
      "127 - HDSigDB_Mouse_2021\n",
      "128 - Jensen_COMPARTMENTS\n",
      "129 - ChEA_2015\n",
      "130 - ChEA_2016\n",
      "131 - KEGG_2019_Mouse\n",
      "132 - ChEA_2022\n",
      "133 - Drug_Perturbations_from_GEO_2014\n",
      "134 - LINCS_L1000_Chem_Pert_Consensus_Sigs\n",
      "135 - TargetScan_microRNA_2017\n",
      "136 - KEGG_2021_Mouse\n",
      "137 - Allen_Brain_Atlas_down\n",
      "138 - WikiPathways_2019_Human\n",
      "139 - Reactome_2013\n",
      "140 - BioCarta_2013\n",
      "141 - Rummagene_transcription_factors\n",
      "142 - Gene_Perturbations_from_GEO_up\n",
      "143 - GO_Cellular_Component_2015\n",
      "144 - Rummagene-signatures\n",
      "145 - GO_Cellular_Component_2013\n",
      "146 - BioCarta_2016\n",
      "147 - Reactome_2015\n",
      "148 - BioCarta_2015\n",
      "149 - Reactome_2016\n",
      "150 - GO_Cellular_Component_2018\n",
      "151 - GO_Cellular_Component_2017\n",
      "152 - GO_Cellular_Component_2023\n",
      "153 - GO_Cellular_Component_2021\n",
      "154 - WikiPathway_2021_Mouse\n",
      "155 - ENCODE_TF_ChIP-seq_2015\n",
      "156 - ENCODE_TF_ChIP-seq_2014\n",
      "157 - RNAseq_Automatic_GEO_Signatures_Mouse_Up\n",
      "158 - GO_Molecular_Function_2017b\n",
      "159 - FANTOM6_lncRNA_KD_DEGs\n",
      "160 - MGI_Mammalian_Phenotype_2013\n",
      "161 - NCI-60_Cancer_Cell_Lines\n",
      "162 - HMS_LINCS_KinomeScan\n",
      "163 - Azimuth_2023\n",
      "164 - MGI_Mammalian_Phenotype_2017\n",
      "165 - Rare_Diseases_GeneRIF_ARCHS4_Predictions\n",
      "166 - Virus_Perturbations_from_GEO_up\n",
      "167 - PFOCR_Pathways\n",
      "168 - SigComLINCS_Drug_Consensus_Signatures\n",
      "169 - IDG_Drug_Targets_2022\n",
      "170 - Enrichr_Libraries_Most_Popular_Genes\n",
      "171 - Orphanet_Augmented_2021\n",
      "172 - GO_Biological_Process_2021\n",
      "173 - TRANSFAC_and_JASPAR_PWMs\n",
      "174 - GO_Biological_Process_2023\n",
      "175 - Rare_Diseases_AutoRIF_ARCHS4_Predictions\n",
      "176 - COVID-19_Related_Gene_Sets\n",
      "177 - Descartes_Cell_Types_and_Tissue_2021\n",
      "178 - Kinase_Perturbations_from_GEO_up\n",
      "179 - Tabula_Muris\n",
      "180 - Tabula_Sapiens\n",
      "181 - Rummagene_kinases\n",
      "182 - TF_Perturbations_Followed_by_Expression\n",
      "183 - GTEx_Aging_Signatures_2021\n",
      "184 - Tissue_Protein_Expression_from_ProteomicsDB\n",
      "185 - DepMap_CRISPR_GeneDependency_CellLines_2023\n",
      "186 - Aging_Perturbations_from_GEO_down\n",
      "187 - GO_Biological_Process_2013\n",
      "188 - GO_Biological_Process_2017b\n",
      "189 - GO_Biological_Process_2018\n",
      "190 - CORUM\n",
      "191 - BioPlex_2017\n",
      "192 - GO_Biological_Process_2015\n",
      "193 - Phosphatase_Substrates_from_DEPOD\n",
      "194 - TRRUST_Transcription_Factors_2019\n",
      "195 - GO_Biological_Process_2017\n",
      "196 - Pfam_InterPro_Domains\n",
      "197 - HuBMAP_ASCT_plus_B_augmented_w_RNAseq_Coexpression\n",
      "198 - Pfam_Domains_2019\n",
      "199 - WikiPathway_2023_Human\n",
      "200 - Allen_Brain_Atlas_up\n",
      "201 - Genome_Browser_PWMs\n",
      "202 - NURSA_Human_Endogenous_Complexome\n",
      "203 - HumanCyc_2015\n",
      "204 - HumanCyc_2016\n",
      "205 - Rummagene_signatures\n",
      "206 - Chromosome_Location_hg19\n",
      "207 - Mouse_Gene_Atlas\n",
      "208 - ChEA_2013\n",
      "209 - miRTarBase_2017\n",
      "210 - GO_Molecular_Function_2023\n",
      "211 - Jensen_DISEASES\n",
      "212 - RNAseq_Automatic_GEO_Signatures_Human_Down\n",
      "213 - Rummagene-transcription-factors\n",
      "214 - ARCHS4_Kinases_Coexp\n",
      "215 - Microbe_Perturbations_from_GEO_down\n",
      "216 - PanglaoDB_Augmented_2021\n",
      "217 - ENCODE_Histone_Modifications_2013\n",
      "218 - Achilles_fitness_decrease\n",
      "219 - ENCODE_Histone_Modifications_2015\n",
      "220 - DSigDB\n",
      "221 - Disease_Perturbations_from_GEO_down\n",
      "222 - DepMap_WG_CRISPR_Screens_Broad_CellLines_2019\n",
      "223 - Drug_Perturbations_from_GEO_down\n",
      "224 - GO_Molecular_Function_2021\n",
      "225 - GO_Molecular_Function_2017\n",
      "226 - GO_Molecular_Function_2018\n",
      "227 - GO_Molecular_Function_2013\n",
      "228 - GO_Molecular_Function_2015\n",
      "229 - Rummagene-kinases\n",
      "230 - TG_GATES_2020\n",
      "231 - KEGG_2021_Human\n",
      "232 - HMDB_Metabolites\n",
      "233 - LINCS_L1000_CRISPR_KO_Consensus_Sigs\n"
     ]
    }
   ],
   "source": [
    "import blitzgsea as blitz\n",
    "# lista de bibliotecas de conjuntos de genes disponibles en Enrichr\n",
    "blitz.enrichr.print_libraries()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "56653505-a4c1-4c5b-9df5-95daa6216140",
   "metadata": {},
   "source": [
    "Bibliotecas escogidas:\n",
    "* Cancer_Cell_Line_Encyclopedia\n",
    "* MSigDB_Computational"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5cdcc5f8-bd6c-4108-a624-8e95e0ece7fe",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Para analizar todas las firma de genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "999819c9-9544-452b-ac95-eaf105ffbfe3",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Introduce el nombre de la librería a utilizar:  MSigDB_Computational\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Error en el archivo 'SLC5A8_2.csv': float division by zero. Saltando este archivo.\n",
      "Proceso completado.\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import blitzgsea as blitz\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "\n",
    "def generar_graficos_desde_excel(result_file_path, signature, library, result):\n",
    "    \"\"\"\n",
    "    Genera y guarda gráficos a partir de un archivo Excel.\n",
    "\n",
    "    Args:\n",
    "        result_file_path (str): Ruta del archivo Excel.\n",
    "        signature: Tu definición de signature.\n",
    "        library: Tu definición de library.\n",
    "        result: Tu definición de result.\n",
    "    \"\"\"\n",
    "    # Leer el archivo .xlsx\n",
    "    df = pd.read_excel(result_file_path)\n",
    "\n",
    "    # Filtrar los términos que cumplen con los criterios especificados\n",
    "    terms = []\n",
    "    for index, row in df.iterrows():\n",
    "        if pd.isna(row[0]) or row[5] > 0.05:\n",
    "            break\n",
    "        terms.append(row[0])\n",
    "\n",
    "    # Carpeta para guardar los resultados\n",
    "    output_folder = file_output_dir\n",
    "\n",
    "    # Generar y guardar los gráficos para cada término\n",
    "    for term in terms:\n",
    "        fig = blitz.plot.running_sum(signature, term, library, result=result, compact=False)\n",
    "        fig.savefig(os.path.join(output_folder, f\"running_sum_{term.replace(' ', '_')}.png\"), bbox_inches='tight')\n",
    "        plt.close(fig)\n",
    "\n",
    "        fig_compact = blitz.plot.running_sum(signature, term, library, result=result, compact=True)\n",
    "        fig_compact.savefig(os.path.join(output_folder, f\"running_sum_compact_{term.replace(' ', '_')}.png\"), bbox_inches='tight')\n",
    "        plt.close(fig_compact)\n",
    "        \n",
    "    # Cerrar todas las figuras abiertas (opcional, por seguridad)\n",
    "    plt.close('all')\n",
    "\n",
    "# Directorios de entrada y salida\n",
    "input_dir = \"Firmas_genes/CSV\"\n",
    "output_dir = \"Resultados\"\n",
    "output_dir2 = \"Hojas_resultados\"\n",
    "\n",
    "# Obtener lista de archivos .csv en el directorio de entrada\n",
    "csv_files = [file for file in os.listdir(input_dir) if file.endswith(\".csv\")]\n",
    "\n",
    "# Solicitar al usuario el nombre de la librería\n",
    "library_name = input(\"Introduce el nombre de la librería a utilizar: \")\n",
    "\n",
    "# Iterar sobre cada archivo .csv en el directorio de entrada\n",
    "for file in csv_files:\n",
    "    # Obtener el nombre del archivo sin la extensión\n",
    "    file_name = os.path.splitext(file)[0]\n",
    "    \n",
    "    # Leer el archivo .csv como pandas DataFrame\n",
    "    signature = pd.read_csv(os.path.join(input_dir, file))\n",
    "    \n",
    "    # Obtener la librería especificada por el usuario\n",
    "    library = blitz.enrichr.get_library(library_name)\n",
    "    \n",
    "    # Realizar análisis de enriquecimiento\n",
    "    try:\n",
    "        if __name__ == \"__main__\":  # make sure process is main, when run in a script it can cause errors otherwise\n",
    "          result = blitz.gsea(signature, library)\n",
    "    except (ZeroDivisionError, ValueError) as e:\n",
    "        print(f\"Error en el archivo '{file}': {e}. Saltando este archivo.\")\n",
    "        continue\n",
    "    \n",
    "    # Crear directorio para los resultados del archivo actual\n",
    "    file_output_dir = os.path.join(output_dir, file_name)\n",
    "    os.makedirs(file_output_dir, exist_ok=True)\n",
    "    \n",
    "    # Guardar los resultados en un archivo .xlsx en el directorio de salida correspondiente\n",
    "    result_file_path = os.path.join(file_output_dir, file_name + \"_resultados_GSEA.xlsx\")\n",
    "    result.to_excel(result_file_path, index=True)\n",
    "    result_file_path2 = os.path.join(output_dir2, file_name + \"_resultados_GSEA.xlsx\")\n",
    "    result.to_excel(result_file_path2, index=True)\n",
    "    \n",
    "    # Generar y guardar el gráfico de la tabla superior como .png\n",
    "    try:\n",
    "        fig_table = blitz.plot.top_table(signature, library, result, n=15)\n",
    "        fig_table.savefig(os.path.join(file_output_dir, \"top_table.png\"), bbox_inches='tight')\n",
    "        plt.close(fig_table)\n",
    "    except IndexError as e:\n",
    "        print(f\"Error al generar el gráfico para el archivo '{file}': {e}. No se guardará el gráfico.\")\n",
    "        \n",
    "    # Llama a la función\n",
    "    generar_graficos_desde_excel(result_file_path, signature, library, result)\n",
    "    \n",
    "    # Cerrar todas las figuras abiertas (opcional, por seguridad)\n",
    "    plt.close('all')\n",
    "    \n",
    "print(\"Proceso completado.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "abcb6299-0f2b-46b9-957a-5e40ca3719da",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "### Para analizar una sola firma de expresión génica"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "be7f7274-8ced-4c00-8d40-5c8d91411ef5",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import blitzgsea as blitz\n",
    "import pandas as pd\n",
    "\n",
    "# read signature as pandas dataframe\n",
    "signature = pd.read_csv(\"Firmas_genes/CSV/APC_2.csv\")\n",
    "\n",
    "# use enrichr submodule to retrieve gene set library\n",
    "library = blitz.enrichr.get_library(\"MSigDB_Computational\")\n",
    "#library = blitz.enrichr.get_library(\"Cancer_Cell_Line_Encyclopedia\")\n",
    "\n",
    "# run enrichment analysis\n",
    "if __name__ == \"__main__\":  # make sure process is main, when run in a script it can cause errors otherwise\n",
    "  result = blitz.gsea(signature, library)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "91024bf2-ecc0-416d-9a53-eb8a00e90282",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>es</th>\n",
       "      <th>nes</th>\n",
       "      <th>pval</th>\n",
       "      <th>sidak</th>\n",
       "      <th>fdr</th>\n",
       "      <th>geneset_size</th>\n",
       "      <th>leading_edge</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Term</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MORF PDPK1</th>\n",
       "      <td>-0.725246</td>\n",
       "      <td>-2.709892</td>\n",
       "      <td>0.006731</td>\n",
       "      <td>0.929633</td>\n",
       "      <td>0.49051</td>\n",
       "      <td>6</td>\n",
       "      <td>PDPK1,LMTK2,ANKRD12,GIGYF2,PHIP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GNF2 MAP2K3</th>\n",
       "      <td>0.749419</td>\n",
       "      <td>2.617741</td>\n",
       "      <td>0.008851</td>\n",
       "      <td>0.969624</td>\n",
       "      <td>0.49051</td>\n",
       "      <td>5</td>\n",
       "      <td>XPO7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MODULE 308</th>\n",
       "      <td>0.752516</td>\n",
       "      <td>2.617267</td>\n",
       "      <td>0.008864</td>\n",
       "      <td>0.969771</td>\n",
       "      <td>0.49051</td>\n",
       "      <td>6</td>\n",
       "      <td>DNMT1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GNF2 PAK2</th>\n",
       "      <td>0.740207</td>\n",
       "      <td>2.576300</td>\n",
       "      <td>0.009986</td>\n",
       "      <td>0.980637</td>\n",
       "      <td>0.49051</td>\n",
       "      <td>5</td>\n",
       "      <td>RAF1,ACAP2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MORF MYC</th>\n",
       "      <td>0.740197</td>\n",
       "      <td>2.561852</td>\n",
       "      <td>0.010412</td>\n",
       "      <td>0.983645</td>\n",
       "      <td>0.49051</td>\n",
       "      <td>6</td>\n",
       "      <td>KAT8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MODULE 163</th>\n",
       "      <td>0.191929</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>17</td>\n",
       "      <td>PGS1,RANBP3,NFATC2IP,FAM199X,FBXL5,GIT2,RBM17,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCM SUFU</th>\n",
       "      <td>0.231214</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>19</td>\n",
       "      <td>FBXO3,NCOA5,RALGAPA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MORF ATF2</th>\n",
       "      <td>0.197500</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>29</td>\n",
       "      <td>MPP3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MORF RFC5</th>\n",
       "      <td>0.233437</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>11</td>\n",
       "      <td>TAF2,SCAMP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MORF ATRX</th>\n",
       "      <td>0.252143</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>36</td>\n",
       "      <td>TAF2,PIK3CB,PHF21A,SCAMP1,KAT8,NMT1,ZNF500</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>393 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   es       nes      pval     sidak      fdr  geneset_size  \\\n",
       "Term                                                                         \n",
       "MORF PDPK1  -0.725246 -2.709892  0.006731  0.929633  0.49051             6   \n",
       "GNF2 MAP2K3  0.749419  2.617741  0.008851  0.969624  0.49051             5   \n",
       "MODULE 308   0.752516  2.617267  0.008864  0.969771  0.49051             6   \n",
       "GNF2 PAK2    0.740207  2.576300  0.009986  0.980637  0.49051             5   \n",
       "MORF MYC     0.740197  2.561852  0.010412  0.983645  0.49051             6   \n",
       "...               ...       ...       ...       ...      ...           ...   \n",
       "MODULE 163   0.191929 -0.000000  1.000000  1.000000  1.00000            17   \n",
       "GCM SUFU     0.231214 -0.000000  1.000000  1.000000  1.00000            19   \n",
       "MORF ATF2    0.197500 -0.000000  1.000000  1.000000  1.00000            29   \n",
       "MORF RFC5    0.233437 -0.000000  1.000000  1.000000  1.00000            11   \n",
       "MORF ATRX    0.252143 -0.000000  1.000000  1.000000  1.00000            36   \n",
       "\n",
       "                                                  leading_edge  \n",
       "Term                                                            \n",
       "MORF PDPK1                     PDPK1,LMTK2,ANKRD12,GIGYF2,PHIP  \n",
       "GNF2 MAP2K3                                               XPO7  \n",
       "MODULE 308                                               DNMT1  \n",
       "GNF2 PAK2                                           RAF1,ACAP2  \n",
       "MORF MYC                                                  KAT8  \n",
       "...                                                        ...  \n",
       "MODULE 163   PGS1,RANBP3,NFATC2IP,FAM199X,FBXL5,GIT2,RBM17,...  \n",
       "GCM SUFU                                  FBXO3,NCOA5,RALGAPA1  \n",
       "MORF ATF2                                                 MPP3  \n",
       "MORF RFC5                                          TAF2,SCAMP1  \n",
       "MORF ATRX           TAF2,PIK3CB,PHF21A,SCAMP1,KAT8,NMT1,ZNF500  \n",
       "\n",
       "[393 rows x 7 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Ejecute este bloque de código si quiere ver un resumen de la tabla resultante de la ejecuión de blitzGSEA.\n",
    "result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "949aa7c2-f1fc-46cc-bee9-cb6beb61b4d2",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "result.to_excel(\"Resultados/Resultados_GSEA.xlsx\", index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f0a0ff02-8152-44d7-b860-ec0fe27ea9d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot the enrichment results and save to pdf\n",
    "fig_table = blitz.plot.top_table(signature, library, result, n=15)\n",
    "fig_table.savefig(\"top_table.png\", bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "90713e35-6dc9-43dd-b23d-701f20dea867",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Leer el archivo .xlsx\n",
    "file_path = 'Resultados/Resultados_GSEA.xlsx'\n",
    "df = pd.read_excel(file_path)\n",
    "\n",
    "# Filtrar los términos que cumplen con los criterios especificados\n",
    "terms = []\n",
    "for index, row in df.iterrows():\n",
    "    if pd.isna(row[0]) or row[5] > 0.49051:\n",
    "        break\n",
    "    terms.append(row[0])\n",
    "\n",
    "# Carpeta para guardar los resultados\n",
    "output_folder = 'Resultados'\n",
    "\n",
    "# Generar y guardar los gráficos para cada término\n",
    "for term in terms:\n",
    "    fig = blitz.plot.running_sum(signature, term, library, result=result, compact=False)\n",
    "    fig.savefig(os.path.join(output_folder, f\"running_sum_{term.replace(' ', '_')}.png\"), bbox_inches='tight')\n",
    "    plt.close(fig)\n",
    "\n",
    "    fig_compact = blitz.plot.running_sum(signature, term, library, result=result, compact=True)\n",
    "    fig_compact.savefig(os.path.join(output_folder, f\"running_sum_compact_{term.replace(' ', '_')}.png\"), bbox_inches='tight')\n",
    "    plt.close(fig_compact)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "07f6e575-392d-4bfb-be6f-0402cd968bf0",
   "metadata": {},
   "source": [
    "## Guardado de los resultados del GSEA estadísticamente significativos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "31540909-001e-43ef-8507-09a3b9e27923",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Los métodos estadísticos disponibles para filtrar los resultados son:\n",
      "1: Tasa de descubrimiento falso (FDR)\n",
      "2: Corrección de Šidák\n",
      "3: La combinación de ambos métodos\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Introduzca el número correspondiente a la opción elegida (1, 2 o 3):  1\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "El archivo IDO2_2_resultados_GSEA.xlsx no contiene términos enriquecidos de forma estadísticamente significativa.\n",
      "El archivo NMRK2_2_resultados_GSEA.xlsx no contiene términos enriquecidos de forma estadísticamente significativa.\n",
      "El archivo Resultados_GSEA.xlsx no contiene términos enriquecidos de forma estadísticamente significativa.\n",
      "El archivo Resultados_GSEA_APCDD1.xlsx no contiene términos enriquecidos de forma estadísticamente significativa.\n",
      "El archivo filtrado ha sido guardado en: Resultados_significativos\\Resultados_GSEA_APC_2_significativos.xlsx\n",
      "Proceso completado.\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "\n",
    "# Ruta de la carpeta que contiene los archivos .xlsx\n",
    "folder_path = 'Hojas_resultados'\n",
    "\n",
    "# Obtener la lista de todos los archivos .xlsx en la carpeta\n",
    "files = [f for f in os.listdir(folder_path) if f.endswith('.xlsx')]\n",
    "\n",
    "# Preguntar al usuario por el método de filtrado\n",
    "print(\"Los métodos estadísticos disponibles para filtrar los resultados son:\")\n",
    "print(\"1: Tasa de descubrimiento falso (FDR)\")\n",
    "print(\"2: Corrección de Šidák\")\n",
    "print(\"3: La combinación de ambos métodos\")\n",
    "filter_choice = input(\"Introduzca el número correspondiente a la opción elegida (1, 2 o 3): \")\n",
    "\n",
    "# Validar la entrada del usuario\n",
    "while filter_choice not in ['1', '2', '3']:\n",
    "    filter_choice = input(\"Entrada no válida. Introduce 1, 2, o 3: \")\n",
    "\n",
    "for file in files:\n",
    "    # Leer el archivo .xlsx\n",
    "    input_path = os.path.join(folder_path, file)\n",
    "    df = pd.read_excel(input_path)\n",
    "\n",
    "    # Filtrar las filas que cumplen las condiciones especificadas\n",
    "    if filter_choice == '1':\n",
    "        filtered_df = df[df['fdr'] <= 0.05]\n",
    "    elif filter_choice == '2':\n",
    "        filtered_df = df[df['sidak'] <= 0.05]\n",
    "    else:\n",
    "        filtered_df = df[(df['fdr'] <= 0.05) & (df['sidak'] <= 0.05)]\n",
    "\n",
    "    # Verificar si el DataFrame filtrado está vacío\n",
    "    if filtered_df.empty:\n",
    "        print(f\"El archivo {file} no contiene términos enriquecidos de forma estadísticamente significativa.\")\n",
    "        continue\n",
    "\n",
    "    # Crear el nombre del archivo de salida\n",
    "    output_file_name = file.replace('.xlsx', '_significativos.xlsx')\n",
    "    output_path = 'Resultados_significativos'\n",
    "    if not os.path.exists(output_path):\n",
    "        os.makedirs(output_path)\n",
    "    output_path = os.path.join(output_path, output_file_name)\n",
    "\n",
    "    # Escribir el DataFrame resultante a un nuevo archivo .xlsx\n",
    "    filtered_df.to_excel(output_path, index=False)\n",
    "\n",
    "    print(f\"El archivo filtrado ha sido guardado en: {output_path}\")\n",
    "    \n",
    "print(\"Proceso completado.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da108f98-ddae-43b5-b65a-0e583ecf8f85",
   "metadata": {},
   "source": [
    "## Filtrar resultados según su relevancia funcional para el tipo de tumor de interés"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "d62887a5-3e2b-4f7d-b837-195a7757255b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No hay filas que cumplan los criterios en el archivo: APCDD1_resultados_GSEA_significativos.xlsx\n",
      "No hay filas que cumplan los criterios en el archivo: APC_2_resultados_GSEA_significativos.xlsx\n",
      "No hay filas que cumplan los criterios en el archivo: Resultados_GSEA_significativos.xlsx\n",
      "Proceso completado.\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "\n",
    "# Definimos las rutas de las carpetas\n",
    "ruta_resultados_significativos = 'Resultados_significativos'\n",
    "ruta_criterios_filtrado = 'Criterios_filtrado'\n",
    "ruta_resultados_filtrados = 'Resultados_filtrados'\n",
    "\n",
    "# Leer los términos de Inclusion.txt y Exclusion.txt\n",
    "with open(os.path.join(ruta_criterios_filtrado, 'Inclusion.txt'), 'r') as file:\n",
    "    inclusion_terms = [line.strip() for line in file]\n",
    "\n",
    "with open(os.path.join(ruta_criterios_filtrado, 'Exclusion.txt'), 'r') as file:\n",
    "    exclusion_terms = [line.strip() for line in file]\n",
    "\n",
    "# Aseguramos que la carpeta de resultados filtrados exista\n",
    "os.makedirs(ruta_resultados_filtrados, exist_ok=True)\n",
    "\n",
    "# Procesar cada archivo .xlsx en la carpeta Resultados_significativos\n",
    "for filename in os.listdir(ruta_resultados_significativos):\n",
    "    if filename.endswith('.xlsx'):\n",
    "        filepath = os.path.join(ruta_resultados_significativos, filename)\n",
    "        df = pd.read_excel(filepath)\n",
    "\n",
    "        # Filtrar las filas según los términos de inclusión y exclusión\n",
    "        inclusion_mask = df['Term'].apply(lambda x: any(term in x for term in inclusion_terms))\n",
    "        exclusion_mask = df['Term'].apply(lambda x: any(term in x for term in exclusion_terms))\n",
    "\n",
    "        filtered_df = df[inclusion_mask & ~exclusion_mask]\n",
    "\n",
    "        if not filtered_df.empty:\n",
    "            # Guardar el nuevo archivo si hay filas que cumplen los criterios\n",
    "            new_filename = filename.replace('_resultados_GSEA_significativos.xlsx', '_resultados_GSEA_RF.xlsx')\n",
    "            filtered_df.to_excel(os.path.join(ruta_resultados_filtrados, new_filename), index=False)\n",
    "        else:\n",
    "            # Avisar al usuario si no hay filas que cumplan los criterios\n",
    "            print(f'No hay filas que cumplan los criterios en el archivo: {filename}')\n",
    "            \n",
    "print(\"Proceso completado.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "027efaac-1e28-4269-97e7-450053088b8b",
   "metadata": {},
   "source": [
    "## Vaciado de las carpetas de archivos\n",
    "Tras copiar los archivos que considere de interés a otra ubicación, puede ejecutar este código para borrar todos los archivos presentes en las carpetas indicadas en la \"Lista de las carpetas\"."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "id": "51613ba8-e10d-4130-a8a7-6b08810e3f8a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Archivo Resultados_significativos\\APCDD1L_resultados_GSEA_significativos.xlsx eliminado\n",
      "Archivo Resultados_significativos\\CAMK2A_resultados_GSEA_significativos.xlsx eliminado\n",
      "Archivo Resultados_significativos\\CSNK1A1_2_resultados_GSEA_significativos.xlsx eliminado\n",
      "Archivo Resultados_significativos\\CSNK2B_resultados_GSEA_significativos.xlsx eliminado\n",
      "Archivo Resultados_significativos\\CTNNBIP1_resultados_GSEA_significativos.xlsx eliminado\n",
      "Archivo Resultados_significativos\\DAAM2_resultados_GSEA_significativos.xlsx eliminado\n",
      "Archivo Resultados_significativos\\DKK2_resultados_GSEA_significativos.xlsx eliminado\n",
      "Archivo Resultados_significativos\\DVL1_2_resultados_GSEA_significativos.xlsx eliminado\n",
      "Archivo Resultados_significativos\\FRZB_resultados_GSEA_significativos.xlsx eliminado\n",
      "Archivo Resultados_significativos\\FZD1_resultados_GSEA_significativos.xlsx eliminado\n",
      "Archivo Resultados_significativos\\NFATC2_resultados_GSEA_significativos.xlsx eliminado\n",
      "Archivo Resultados_significativos\\NFATC4_resultados_GSEA_significativos.xlsx eliminado\n",
      "Archivo Resultados_significativos\\PLCB2_resultados_GSEA_significativos.xlsx eliminado\n",
      "Archivo Resultados_significativos\\PRICKLE4_resultados_GSEA_significativos.xlsx eliminado\n",
      "Archivo Resultados_significativos\\PRKCA_resultados_GSEA_significativos.xlsx eliminado\n",
      "Archivo Resultados_significativos\\PRKCB_resultados_GSEA_significativos.xlsx eliminado\n",
      "Archivo Resultados_significativos\\ROCK2_2_resultados_GSEA_significativos.xlsx eliminado\n",
      "Archivo Resultados_significativos\\RSPO1_resultados_GSEA_significativos.xlsx eliminado\n",
      "Archivo Resultados_significativos\\RSPO3_resultados_GSEA_significativos.xlsx eliminado\n",
      "Archivo Resultados_significativos\\SERPINF1_resultados_GSEA_significativos.xlsx eliminado\n",
      "Archivo Resultados_significativos\\SFRP2_resultados_GSEA_significativos.xlsx eliminado\n",
      "Archivo Resultados_significativos\\SFRP4_resultados_GSEA_significativos.xlsx eliminado\n",
      "Archivo Resultados_significativos\\SOX17_resultados_GSEA_significativos.xlsx eliminado\n",
      "Archivo Resultados_significativos\\WNT2_resultados_GSEA_significativos.xlsx eliminado\n",
      "Proceso completado.\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "# Lista de las carpetas\n",
    "carpetas = [\n",
    "    \"Hojas_resultados\",\n",
    "    \"Resultados_significativos\",\n",
    "    \"Resultados_filtrados\"\n",
    "]\n",
    "\n",
    "def eliminar_archivos(carpeta):\n",
    "    # Listar todos los elementos en la carpeta\n",
    "    for filename in os.listdir(carpeta):\n",
    "        file_path = os.path.join(carpeta, filename)\n",
    "        try:\n",
    "            # Verificar si es un archivo y eliminarlo\n",
    "            if os.path.isfile(file_path) or os.path.islink(file_path):\n",
    "                os.unlink(file_path)\n",
    "                print(f'Archivo {file_path} eliminado')\n",
    "            # Si es un directorio, lo ignoramos\n",
    "        except Exception as e:\n",
    "            print(f'No se pudo eliminar {file_path}. Razón: {e}')\n",
    "\n",
    "# Eliminar archivos en cada carpeta\n",
    "for carpeta in carpetas:\n",
    "    eliminar_archivos(carpeta)\n",
    "    \n",
    "print(\"Proceso completado.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c7df0bb-6648-40e4-94d1-6b46019e8e8c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
